src/flow/src/compute/render/reduce.rs

```diff
@@ -0,0 +1,957 @@
+///
+/// invariant: it'is assumed `kv`'s time is always <= now,
+/// since it's from a Collection Bundle, where future inserts are stored in arrange
+fn reduce_distinct_subgraph(
+    arrange: &ArrangeHandler,
+    kv: impl IntoIterator<Item = KeyValDiffRow>,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler: _,
+        send,
+    }: SubgraphArg,
+) {
+    let ret = update_reduce_distinct_arrange(arrange, kv, now, err_collector).collect_vec();
+
+    // no future updates should exist here
+    if arrange.read().get_next_update_time(&now).is_some() {
+        err_collector.push_err(
+            InternalSnafu {
+                reason: "No future updates should exist in the reduce distinct arrangement",
+            }
+            .build(),
+        );
+    }
+
+    send.give(ret);
+}
+
+/// eval accumulable reduce plan by eval aggregate function and reduce the result
+///
+/// TODO: eval distinct by adding distinct input arrangement
+///
+/// invariant: it'is assumed `kv`'s time is always <= now,
+/// since it's from a Collection Bundle, where future inserts are stored in arrange
+fn reduce_accum_subgraph(
+    arrange: &ArrangeHandler,
+    distinct_input: &Option<Vec<ArrangeHandler>>,
+    kv: impl IntoIterator<Item = KeyValDiffRow>,
+    accum_plan: &AccumulablePlan,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler,
+        send,
+    }: SubgraphArg,
+) {
+    let AccumulablePlan {
+        full_aggrs,
+        simple_aggrs,
+        distinct_aggrs,
+    } = accum_plan;
+    let mut key2vals = BTreeMap::<Row, Vec<(Row, repr::Diff)>>::new();
+
+    for ((key, val), _tick, diff) in kv {
+        // it is assumed that value is in order of insertion
+        let vals = key2vals.entry(key).or_default();
+        vals.push((val, diff));
+    }
+
+    let mut all_updates = Vec::with_capacity(key2vals.len());
+    let mut all_outputs = Vec::with_capacity(key2vals.len());
+
+    // lock the arrange for write
+    let mut arrange = arrange.write();
+    for (key, value_diffs) in key2vals {
+        let col_diffs = {
+            let row_len = value_diffs[0].0.len();
+            // split value_diffs into columns
+            (0..row_len)
+                .map(|i| {
+                    value_diffs
+                        .iter()
+                        .map(|(row, diff)| (row.get(i).cloned().unwrap(), *diff))
+                        .collect_vec()
+                })
+                .collect_vec()
+        };
+        let (accums, _, _) = arrange.get(now, &key).unwrap_or_default();
+        let accums = accums.inner;
+
+        // deser accums from offsets
+        let accum_ranges = {
+            let res = err_collector
+                .run(|| from_val_to_slice_idx(accums.first().cloned(), full_aggrs.len()));
+            if let Some(res) = res {
+                res
+            } else {
+                // ignore for whatever reason
```

Comment:
I'm not certain whether we need to add debug logs for this specific corner case. It seems like it could be challenging to troubleshoot runtime flow tasks if we encounter issues. Additionally, it appears that the flow crate lacks sufficient logs.

Reply:
> I'm not certain whether we need to add debug logs for this specific corner case. It seems like it could be challenging to troubleshoot runtime flow tasks if we encounter issues. Additionally, it appears that the flow crate lacks sufficient logs.

All errors generated in flow's eval process is collected in `err_collector`, and the logging of `err_collector` should be sufficient?(although not yet impl), as err_collector is now per-dataflow, might add per-subgraph location later

Reply:
Got it. That is enough.